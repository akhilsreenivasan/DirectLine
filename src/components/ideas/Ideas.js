import Challenges from "./challenges/challenges";
import ideacss from "./ideas.css";
import { useEffect, useState } from "react";
import Button from "@material-ui/core/Button";
import { useDispatch, useSelector } from "react-redux";
const Ideas = (props) => {
  const [toggle, setToggle] = useState(true);
  const [sortByVote, setSortByVote] = useState(true);
  const [sortByDate, setSortByDate] = useState(true);
  const [hackdata, setHackdata] = useState([]);
  const [sortedData, setsortedData] = useState([]);
  const [lastSort, setlastSort] = useState();
  const dispatch = useDispatch();
  const allHacks = useSelector((state) => state.hacks);
  const { user } = props;
  useEffect(() => {
    let data = JSON.parse(localStorage.getItem("AllHacks")) || [];
    dispatch({ type: "GET_ALL_HACKS", data });
  }, []);
  useEffect(() => {
    setHackdata(allHacks);
    if (toggle) {
      if (lastSort?.name == "vote") {
        sortVote(lastSort.flag, allHacks);
      } else if (lastSort?.date == "vote") {
        sortDate(lastSort.flag, allHacks);
      } else {
        setsortedData(allHacks);
      }
    } else {
      const filtered = allHacks.filter((ch) => ch.userId == user.id);
      if (lastSort?.name == "vote") {
        sortVote(lastSort.flag, filtered);
      } else if (lastSort?.date == "vote") {
        sortDate(lastSort.flag, filtered);
      } else {
        setsortedData(filtered);
      }
    }
  }, [allHacks]);
  const sortByDateFn = (flag) => {
    setSortByDate((prev) => {
      return !prev;
    });
    setlastSort({ name: "date", flag });
    sortDate(flag, sortedData);
  };
  const sortDate = (flag, dataset) => {
    let data = [...dataset];
    if (flag) {
      data.sort((a, b) => a.creationDate - b.creationDate);
    } else {
      data.sort((a, b) => b.creationDate - a.creationDate);
    }
    setsortedData(data);
  };
  const sortByVoteFn = (flag) => {
    setSortByVote((prev) => {
      return !prev;
    });
    setlastSort({ name: "vote", flag });
    sortVote(flag, sortedData);
  };
  const sortVote = (flag, datset) => {
    let data = [...datset];
    if (flag) {
      data.sort((a, b) => a.votedList.length - b.votedList.length);
    } else {
      data.sort((a, b) => b.votedList.length - a.votedList.length);
    }
    setsortedData(data);
  };
  const toggleData = (flag) => {
    setToggle((prev) => {
      return !prev;
    });
    if (!flag) {
      if (lastSort?.name == "vote") {
        sortVote(lastSort.flag, allHacks);
      } else if (lastSort?.date == "vote") {
        sortDate(lastSort.flag, allHacks);
      } else {
        setsortedData(allHacks);
      }
    } else {
      const filtered = hackdata.filter((ch) => ch.userId == user.id);
      if (lastSort?.name == "vote") {
        sortVote(lastSort.flag, filtered);
      } else if (lastSort?.date == "vote") {
        sortDate(lastSort.flag, filtered);
      } else {
        setsortedData(filtered);
      }
    }
  };
  const upVote = (challenge, vote) => {
    dispatch({ type: "ADD_VOTE", data: { challenge, vote, user } });
  };
  return (
    <div className={ideacss.challenges}>
      <h2>{toggle ? "All Challenges" : "My Challenges"}</h2>
      <Button
        size="small"
        data-testid="toggle-challenges"
        color="secondary"
        disabled={false}
        variant="outlined"
        style={{ marginTop: "15px" }}
        onClick={toggleData.bind(this, toggle)}
      >
        {!toggle ? "View All Challenges" : "View My Challenges"}
      </Button>
      <Button
        disabled={!sortedData.length}
        color="primary"
        size="small"
        variant="contained"
        style={{ marginTop: "15px", marginLeft: "15px" }}
        onClick={sortByDateFn.bind(this, sortByDate)}
      >
        {sortByDate ? "SortBy Date (ASC)" : "SortBy Date (DSC)"}
      </Button>
      <Button
        disabled={!sortedData.length}
        color="primary"
        size="small"
        variant="contained"
        style={{ marginTop: "15px", marginLeft: "15px" }}
        onClick={sortByVoteFn.bind(this, sortByVote)}
      >
        {sortByVote ? "SortBy Vote (ASC)" : "SortBy Vote (DSC)"}
      </Button>
      {!sortedData.length ? (
        <p style={{ textAlign: "center", marginTop: "90px" }}>
          No Challenges Available
        </p>
      ) : (
        ""
      )}
      {
        <Challenges
          hackdata={sortedData}
          user={user}
          upVote={upVote}
        ></Challenges>
      }
    </div>
  );
};

export default Ideas;
