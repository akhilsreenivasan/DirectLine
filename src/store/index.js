import { createStore } from "redux";
export const initialState = {
    user: [],
    hacks: []
}
export const hackIdeas = (state = initialState, action) => {
    let { hacks } = state;
    switch (action.type) {
        case 'SAVE_USER_DATA':
            return Object.assign({}, state, { user: action.data });
        case 'CLEAR_USER_DATA':
            return Object.assign({}, state, { user: [] });
        case 'GET_ALL_HACKS':
            return Object.assign({}, state, { hacks: action.data });
        case 'ADD_NEW_CHALLENGE':
            localStorage.setItem('AllHacks', JSON.stringify([...hacks, action.data]));
            return Object.assign({}, state, { hacks: [...hacks, action.data] });
        case 'ADD_VOTE':
            let data = [...hacks];
            let index = data.indexOf(action.data.challenge);
            let { user } = action.data;
            if (!action.data.vote) {
                data[index].votedList.push(user.id);
                localStorage.setItem('AllHacks', JSON.stringify(data));
                return Object.assign({}, state, { hacks: data });
            } else {
                let index1 = data[index].votedList.indexOf(user.id);
                data[index].votedList.splice(index1, 1);
                localStorage.setItem('AllHacks', JSON.stringify(data));
                return Object.assign({}, state, { hacks: data });
            }
    }
    return state;
}
const store = createStore(hackIdeas);

export default store;