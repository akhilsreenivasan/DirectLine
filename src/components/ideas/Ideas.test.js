import * as react from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import Ideas from "./Ideas";
import { hackIdeas, initialState } from "../../store";
import { createStore } from "redux";
import { useDispatch, useSelector } from "react-redux";
function renderWithProviders(ui, { reduxState } = {}) {
  const store = createStore(hackIdeas, reduxState || initialState);
  return render(<Provider store={store}>{ui}</Provider>);
}
describe("submit employee id", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  const user = { name: "Joe", id: 1500 };
  test("Test Initial Rendering Without Data", () => {
    let user = { name: "Joe", id: 1500 };
    let hacks = [];
    const { getByText, container } = renderWithProviders(
      <Ideas user={user} />,
      {
        reduxState: { user, hacks },
      }
    );
    waitFor(() => {
      expect(getByText(/ChallengesCmp/i)).toBeInTheDocument();
    });
    getByText(/All Challenges/i);
    getByText(/No Challenges Available/i);
    expect(container).toMatchSnapshot();
  });

  test("Test Initial Rendering With Data", () => {
    let user = { name: "Joe", id: 1501 };
    let hacks = [
      {
        id: 1,
        title: "One",
        description:
          "Test Description Test Description Test Description Test Description Test Description Test Description Test Description Test Description",
        tags: ["name", "two", "three"],
        userId: 1501,
        creationDate: 1645266448714,
        userName: "Joe",
        votedList: [1500, 1503],
      },
    ];
    const { getByText, container } = renderWithProviders(
      <Ideas user={user} />,
      {
        reduxState: { user, hacks },
      }
    );
    getByText(/All Challenges/i);
    getByText(/No Challenges Available/i);
    expect(container).toMatchSnapshot();
  });

  test("Toggle changes button", () => {
    let user = { name: "Joe", id: 1500 };
    let hacks = [
      {
        id: 1,
        title: "One",
        description:
          "Test Description Test Description Test Description Test Description Test Description Test Description Test Description Test Description",
        tags: ["name", "two", "three"],
        userId: 1500,
        creationDate: 1645266448714,
        userName: "Joe",
        votedList: [1500, 1503],
      },
    ];
    const { getByText, getByTestId, container } = renderWithProviders(
      <Ideas user={user} />,
      {
        reduxState: { user, hacks },
      }
    );
    const button = getByTestId("toggle-challenges");
    fireEvent.click(button);
    waitFor(() => {
      expect(getByText(/No Challenges Available/i)).toBeInTheDocument();
    });
    expect(getByText(/My Challenges/i)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
