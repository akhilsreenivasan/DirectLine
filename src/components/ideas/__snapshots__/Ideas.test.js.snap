// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`submit employee id Test Initial Rendering With Data 1`] = `
<div>
  <div>
    <h2>
      All Challenges
    </h2>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
      data-testid="toggle-challenges"
      style="margin-top: 15px;"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        View My Challenges
      </span>
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall Mui-disabled Mui-disabled"
      disabled=""
      style="margin-top: 15px; margin-left: 15px;"
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        SortBy Date (ASC)
      </span>
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall Mui-disabled Mui-disabled"
      disabled=""
      style="margin-top: 15px; margin-left: 15px;"
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        SortBy Vote (ASC)
      </span>
    </button>
    <p
      style="text-align: center; margin-top: 90px;"
    >
      No Challenges Available
    </p>
    <div
      class="challenge"
    />
  </div>
</div>
`;

exports[`submit employee id Test Initial Rendering Without Data 1`] = `
<div>
  <div>
    <h2>
      All Challenges
    </h2>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
      data-testid="toggle-challenges"
      style="margin-top: 15px;"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        View My Challenges
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall Mui-disabled Mui-disabled"
      disabled=""
      style="margin-top: 15px; margin-left: 15px;"
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        SortBy Date (ASC)
      </span>
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall Mui-disabled Mui-disabled"
      disabled=""
      style="margin-top: 15px; margin-left: 15px;"
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        SortBy Vote (ASC)
      </span>
    </button>
    <p
      style="text-align: center; margin-top: 90px;"
    >
      No Challenges Available
    </p>
    <div
      class="challenge"
    />
  </div>
</div>
`;

exports[`submit employee id Toggle changes button 1`] = `
<div>
  <div>
    <h2>
      My Challenges
    </h2>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
      data-testid="toggle-challenges"
      style="margin-top: 15px;"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        View All Challenges
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall Mui-disabled Mui-disabled"
      disabled=""
      style="margin-top: 15px; margin-left: 15px;"
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        SortBy Date (ASC)
      </span>
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall Mui-disabled Mui-disabled"
      disabled=""
      style="margin-top: 15px; margin-left: 15px;"
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        SortBy Vote (ASC)
      </span>
    </button>
    <p
      style="text-align: center; margin-top: 90px;"
    >
      No Challenges Available
    </p>
    <div
      class="challenge"
    />
  </div>
</div>
`;
